baseapp/baseapp.go

```diff
@@ -696,26 +698,57 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 		if app.postHandler != nil {
 			newCtx, err := app.postHandler(runMsgCtx, tx, mode == runTxModeSimulate)
 			if err != nil {
-				return gInfo, nil, nil, priority, err
+				return gInfo, nil, nil, priority, ctx, err
 			}
 
 			result.Events = append(result.Events, newCtx.EventManager().ABCIEvents()...)
 		}
 
 		if mode == runTxModeDeliver {
 			// When block gas exceeds, it'll panic and won't commit the cached store.
+			// this should be called before we charge additional fee( otherwise would
+			// defy the whole point of charging additional fee at the end)
 			consumeBlockGas()
 
 			msCache.Write()
 		}
 
-		if len(anteEvents) > 0 && (mode == runTxModeDeliver || mode == runTxModeSimulate) {
-			// append the events in the order of occurrence
-			result.Events = append(anteEvents, result.Events...)
+		// apply fee logic calls
+		feeEvents, errFromFeeInvoker := FeeInvoke(mode, app, runMsgCtx)
+		// if err from FeeInvoke then don't write to cache
+		if errFromFeeInvoker == nil {
+			msCache.Write()
+			// these are the ante events propagated only on success, that now means that fee charging has happened successfully.
+			if mode == runTxModeDeliver || mode == runTxModeSimulate {
+				if len(anteEvents) > 0 {
+					// append the events in the order of occurrence
+					result.Events = append(anteEvents, result.Events...)
+				}
+				// additional fee events
+				if len(feeEvents) > 0 {
+					// append the fee events at the end of the other events, since they get charged at the end of the Tx
+					result.Events = append(result.Events, feeEvents.ToABCIEvents()...)
+				}
+			}
```

Comment:
I'm not fully sure if this should be this way, or if the fee events should always be added (regardless of mode).

E.g.
```golang
if len(anteEvents) > 0 && (mode == runTxModeDeliver || mode == runTxModeSimulate) {
        // append the events in the order of occurrence
        result.Events = append(anteEvents, result.Events...)
}
// additional fee events
if len(feeEvents) > 0 {
        // append the fee events at the end of the other events, since they get charged at the end of the Tx
        result.Events = append(result.Events, feeEvents.ToABCIEvents()...)
}
```

Reply:
I believe we should keep it as is.  We tried to keep the flow as close to the sdk as possible.

Reply:
Sounds good! I kept it.